ggshield/utils/git_shell.py

```diff
@@ -265,6 +285,28 @@ def get_last_commit_sha_of_branch(branch_name: str) -> Optional[str]:
     return last_target_commit[0]
 
 
+def get_repository_url_from_path(wd: Path) -> Optional[str]:
+    """
+    Returns one of the repository remote urls. Returns None if no remote are found,
+    or the directory is not a repository.
+    """
+    try:
+        if not is_git_dir(wd):
+            return None
+        remotes_raw = git(["remote", "-v"], cwd=wd).splitlines()
+        matches = (re.search(r"^.*\t(.*) \(.*\)$", remote) for remote in remotes_raw)
+        return next(
+            (
+                _parse_from_git_url(match.group(1))
+                for match in matches
+                if match is not None and len(match.groups()) >= 1
+            ),
+            None,
+        )
+    except Exception:
+        return None
```

Comment:
A few remarks here:

I would prefer if the whole function were not wrapped in a `try..except`. Just wrap what can go wrong, and use more precise exception classes. This avoids hiding bugs.

The code is a bit convoluted, it could be simplified to this:

```python
    for line in git(["remote", "-v"]).splitlines():
        if match := re.search(r"^.*\t(.*) \(.*\)$", line):
            return _parse_from_git_url(match.group(1))
```

But neither the loop in the PR or that proposal prefer the `origin` remote, which is surprising. This version uses `origin` if available:

```python
    url: Optional[str] = None
    for line in git(["remote", "-v"]).splitlines():
        if match := re.search(r"^(.*)\t(.*) \(fetch\)$", line):
            name, url = match.groups()
            if name == "origin":
                break
    return _parse_from_git_url(url) if url else None
```

Reply:
Thank you for this level of detail.

As this change impacts not only iac, but also sca and secrets, we do not want to break existing features of other verticals in case something goes wrong, hence the big try/except. wdyt @GG-HH, can we remove it here?

Also I agree `origin` should have the priority. @GG-HH ?

Reply:
I agree too for the priority of "origin".

For the try/except, I think we should keep it for the git command. For the matching, we should be confident enough about our code and be able to remove it.